src/Storages/Distributed/DirectoryMonitor.cpp

```diff
@@ -715,28 +731,45 @@ void StorageDistributedDirectoryMonitor::processFilesWithBatching(const std::map
         size_t total_rows = 0;
         size_t total_bytes = 0;
         Block sample_block;
-        Settings insert_settings;
-        String insert_query;
-        ClientInfo client_info;
+        DistributedHeader header;
         try
         {
             /// Determine metadata of the current file and check if it is not broken.
             ReadBufferFromFile in{file_path};
-            readHeader(in, insert_settings, insert_query, client_info, log);
-
-            CompressedReadBuffer decompressing_in(in);
-            NativeBlockInputStream block_in(decompressing_in, DBMS_TCP_PROTOCOL_VERSION);
-            block_in.readPrefix();
+            header = readDistributedHeader(in, log);
 
-            while (Block block = block_in.read())
+            if (header.rows)
             {
-                total_rows += block.rows();
-                total_bytes += block.bytes();
+                total_rows += header.rows;
+                total_bytes += header.bytes;
 
-                if (!sample_block)
+                CompressedReadBuffer decompressing_in(in);
+                NativeBlockInputStream block_in(decompressing_in, DBMS_TCP_PROTOCOL_VERSION);
+                block_in.readPrefix();
+
+                /// We still need to read one block for the header.
+                while (Block block = block_in.read())
+                {
                     sample_block = block.cloneEmpty();
+                    break;
+                }
```

Comment:
In c939d2f96ee71beaffbf41429080078c3d206ece Distributed INSERTs was splitted by block structure, and this requires reading the whole file (due to the nature of NativeBlockInputStream)

@alexey-milovidov maybe you can provide some details for CLICKHOUSE-3163 (it was referenced in the mentioned commit)?

Reply:
Actually maybe it was required before #10135 and since it was merged I don't see any reason for doing this

Reply:
Nevermind, this is still relevant, since the underlying table can be modified while distributed not, and in this case the whole batch will fail, need better approach to avoid extra file reading.

Reply:
Should be addressed in ecae6c1c609fbabaf940ec41c40c7f3b2401b412

Reply:
And let's cover this behavior - #19155